### ISSUE TITLE
padding for nn.AvgPool3d?

### ISSUE THREAD

**Issue by thuyen** [pytorch/pytorch#915]
Is there any reason for, among all pooling operations, only `nn.AvgPool3d` doesn't have `padding` argument?

**Comment by soumith** [comment]
there isn't a particular reason, we haven't implemented it yet in our C backend.

**Comment by soumith** [comment]
Fleshing this task out for community members to pick up (Jeyte Hagaley).

This is a task in which you will be modifying the neural network layer `nn.AvgPool3d` to have implicit zero padding.

To visually explain, this is a layer that does Average downsampling, like this:
https://gist.github.com/anonymous/26ebfabaded1c3ee8a7cf0c63e305737

Now, while `nn.AvgPool2d` has support for padding, `nn.AvgPool3d` does not.
You will be adding this feature.

You will look at the nn.AvgPool2d code and understand how padding is introduced in that code, and replicate the same behavior in 3d.
Here are the code pointers:

1. Python entry point: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/pooling.py#L553
2. Calls here: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L203
3. Calls here: https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/pooling.py#L340
4. Now we are entering C
5. The CPU version: https://github.com/pytorch/pytorch/blob/master/torch/lib/THNN/generic/VolumetricAveragePooling.c
6. The GPU version: https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/generic/VolumetricAveragePooling.cu
7. Both CPU and GPU version have to be modified. 

Start with the CPU version and write unit tests in test_nn.py (search for existing tests).
Once you finish the CPU version, modify the GPU version VolumetricAveragePooling.cu . If you do not have a NVIDIA CUDA based GPU I will take over the PR from here to test the GPU side of things.

You will have more questions, and we shall resolve them via comments here or on slack.

Lastly, look at the [CONTRIBUTING](https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#developing-locally-with-pytorch) document for tips on locally developing pytorch.

**Comment by thuyen** [comment]
@soumith I will take care of this. Can you add me on Slack? My email: thuyen@ece.ucsb.edu

**Comment by soumith** [comment]
hi thuyen, jeyte is already working on this. if you are interested i can give you another task

**Comment by thuyen** [comment]
Sorry I didn't read your comment carefully. I would love to contribute, but probably only something in line with what I am currently working on :)

**Comment by jahagaley** [comment]
Just made first pull request. Updated python files, now working on C code.

**Comment by soumith** [comment]
excellent, looking forward to the C code.

**Comment by soumith** [comment]
@jahagaley how is the progress, do you have any questions?

**Comment by jahagaley** [comment]
Almost done. I will make the last pull request by this weekend with my finished C code. I do not have a NVIDIA GPU, is there any way I can get around that?

**Comment by soumith** [comment]
great. thanks for the update.
For now you can push the CPU version, since you dont have NVIDIA GPU, I will do the CUDA part myself.

**Comment by nom** [comment]
@soumith How is the progress with the CUDA part?

**Comment by soumith** [comment]
fixed via https://github.com/pytorch/pytorch/pull/2291
